### ISSUE TITLE
GDAL does not get subdatasets for a WMS service

### ISSUE THREAD

**Issue by michkowalczuk** [OSGeo/gdal#10922]
### What is the bug?

GDAL does not return any dataset for the following WMS service:
`https://mapy.geoportal.gov.pl/wss/service/PZGIK/BDOO/WMS/aktualne`
The service contains 45 layers.

### Steps to reproduce the issue

`gdalinfo "WMS:https://mapy.geoportal.gov.pl/wss/service/PZGIK/BDOO/WMS/aktualne?service=wms&request=getcapabilities"`

### Versions and provenance

Windows 11 / GDAL 3.9.2, released 2024/08/13 from Conda-Forge

### Additional context

Perhaps this issue is also a good moment for refactor the process of getting subdatasets from WMS services, as some of us have been discussing on the [gdal mailing list](https://www.mail-archive.com/gdal-dev@lists.osgeo.org/msg41499.html).

Current GetSubdataset implementation for WMS returns the path only to one combination of layer-SRS, where WMS service provides individual list of CRS for each layer and arbitrarily selected SRS is not usually the best (default) SRS for the presented area.

**Comment by jratike80** [comment]
Maybe it should work also as you did, but by defining version 1.3.0 GDAL finds the layers.

`gdalinfo "WMS:https://mapy.geoportal.gov.pl/wss/service/PZGIK/BDOO/WMS/aktualne?service=wms&version=1.3.0&request=getcapabilities"`

**Comment by michkowalczuk** [comment]
> Maybe it should work also as you did, but by defining version 1.3.0 GDAL finds the layers.

OK, but <WMS_Capabilities> tag contains information about service version.
`<WMS_Capabilities xmlns="http://www.opengis.net/wms" ... version="1.3.0" ...>`
So this information should be used by GDAL when connecting with server, and not transferring the responsibility of including the version in the URL by the user...

**Comment by rouault** [comment]
Tempative fix/workaround for this buggy server in https://github.com/OSGeo/gdal/pull/10924 (not sure that things works end to end as the server only seems to return blank data, at least when working with the smallest overview)

> Perhaps this issue is also a good moment for refactor the process of getting subdatasets from WMS services, as some of us have been discussing on the [gdal mailing list](https://www.mail-archive.com/gdal-dev@lists.osgeo.org/msg41499.html).

You're welcome to work on that.

**Comment by jratike80** [comment]
From the documentation https://gdal.org/en/latest/drivers/raster/wms.html

> WMS version. (optional, defaults to 1.1.1)

GDAL behaves in a documented way and makes this request:
https://mapy.geoportal.gov.pl/wss/service/PZGIK/BDOO/WMS/aktualne?SERVICE=WMS&REQUEST=GetCapabilities&VERSION=1.1.1

The request does get a WMS 1.1.1 GetCapabilities document back. By a quick look I can't say why finding  of the layers fails. Either there is a bug in GDAL or the XML response from the server is invalid. This is the real issue.

It is true that by the WMS standard `&version=` is not mandatory in GetCapabilities. However, version is mandatory in all other WMS requests, and there are servers which think that it is also mandatory in GetCapabilities. For my mind it is more reliable to define the version explicitly because then I know what version will be used.
Note 1. That server gives nice exceptions https://mapy.geoportal.gov.pl/wss/service/PZGIK/BDOO/WMS/aktualne?service=wms&request=getmap
Note 2. A standard complient WMS server should give an exception with https://mapy.geoportal.gov.pl/wss/service/PZGIK/BDOO/WMS/aktualne?service=wms&request=getcapabilities
The values in key-value pairs should be case-sensitive and they should be `WMS` and `GetCapabilities`. GDAL makes that correction automatically, even this server seems to be lenient.

**Comment by rouault** [comment]
> By a quick look I can't say why finding of the layers fails.

the server uses the "CRS" XML element, which is a WMS 1.3.0 thing. WMS 1.1.1 should use "SRS" instead.

**Comment by jratike80** [comment]
> Tempative fix/workaround for this buggy server in https://github.com/OSGeo/gdal/pull/10924

I am against that fix. @michkowalczuk can rather contact the service provider and ask them to fix their WMS 1.1.1 GetCapabilities, or disable the version if they do not want to support it.

**Comment by jratike80** [comment]
The pseudo WFS 1.1.1 document is rubbish also in other places. All this belongs to WMS 1.3.0.

```
<EX_GeographicBoundingBox>
				<westBoundLongitude>11.552492</westBoundLongitude>
				<eastBoundLongitude>28.273838</eastBoundLongitude>
				<southBoundLatitude>48.436556</southBoundLatitude>
				<northBoundLatitude>56.293252</northBoundLatitude>
			</EX_GeographicBoundingBox>
			<BoundingBox CRS="CRS:84" minx="11.552492" miny="48.436556" maxx="28.273838" maxy="56.293252"/>
			<BoundingBox CRS="EPSG:4326" minx="48.436556" miny="11.552492" maxx="56.293252" maxy="28.273838"/>
			<BoundingBox CRS="EPSG:2180" minx="92823.634163" miny="36968.505225" maxx="936846.155542" maxy="1078106.004166"/>
```

**Comment by jratike80** [comment]
An excerpt from the WMS 1.3.0 standard to clarify that it is OK to use version like GDAL does (though it is not mandatory) , and sending back an invalid WMS 1.1.1 document is definitely wrong.

> In response to a GetCapabilities request (for which the VERSION parameter is optional) 
> that does not specify a version number, the server shall respond with the highest version it supports. In response to a GetCapabilities request containing a version number that the server implements, the server shall send that version. If the server does not support the requested version, the server shall respond with output that conforms to a version it does support

**Comment by michkowalczuk** [comment]
> I am against that fix. @michkowalczuk can rather contact the service provider and ask them to fix their WMS 1.1.1 GetCapabilities, or disable the version if they do not want to support it.

Yes, I have just informed them.

> From the documentation https://gdal.org/en/latest/drivers/raster/wms.html
> > WMS version. (optional, defaults to 1.1.1)

I wonder if this assumption is still the best solution? Why not to use the default version that is returned by server in GetCapabilities?

**Comment by jratike80** [comment]
> I wonder if this assumption is still the best solution? Why not to use by default version that is returned by server in GetCapabilities first?

Using the default value (that is always the highest that the server supports) might work if all WMS servers nowadays follow the standard and do not require the `&VERSION=` parameter. That was not the case a few years ago but maybe it has been changed. It took also a decade or so before the different WMS 1.3.0 servers inplemented the order of coordinates in the BBOX with different coordinate systems in a similar way. Using WMS 1.1.1 prevented that mess, and WMS 1.3.0 is from users point of view not any better of different than WMS 1.1.1. The SRS was changed into CRS and the bbox must be written sometimes in another order, but the maps are the same and WMS 1.1.1 had much less issues. But I know that the WMS 1.3.0 support has become better during the time.

I think that the advantage of leaving out the &VERSION= from WMS GetCapabilities request is rather nominal. It would not make big trouble either. Mostly when something fails user needs to do a few more checks for knowing what version gets really used.

> By the way, does GDAL somehow negotiate the service version with the server?

No, I do not believe so. OGC has made many trials with version negotiation in different services (WMS, WFS, WCS). In the latest take of version negotiation OGC had noticed that GetCapabilities without version did not work well and then they introduced AcceptVersions. This is from the OGC Common 1.1.0 standard:

> Version negotiation is performed using the optional AcceptVersions parameter in the 
> GetCapabilities operation request. Although optional, client software should always 
> include this parameter, to simplify version negotiation. The value of this parameter is a 
> sequence of protocol version numbers that the client supports, in order of client 
> preference.

So, effectively they recommend the same that GDAL is doing already: tell which version you want.

WMS 1.3.0 is older that AcceptVersions and it is not supported. I think that only WFS 2.0 and WCS 2.0 support AcceptVersions.
